### DISCUSSION TITLE
Genealogical Relatedness in pyslim

### DISCUSSION THREAD

**Discussion by fouerghi** [tskit-dev/pyslim#325] [Q&A] [answered]
Hi everyone, I am currently building a simulation where I would like to track the ancestry all the way back to the first generation. SLiM provides the relatedness feature that allows to get 2 generations back for each individual. However, I am interested in 10 generations back. That's why I am creating a tree sequence. However, I am stuck on the part in pyslim, where I can calculate the genealogical relatedness. I just want to use the pedigree to calculate what the relatedness() feature does in SLiM but all the way back to the first generation. Does anyone have advice on how to do this in pyslim?

**Comment by petrelharp** [comment, marked as answer]
Hello, @fouerghi! I do have advice. Let's see: It depends *which* relatedness you are interested in. If it is the "realized" relatedness (ie. the amount of genetic material inherited in common), you can use [ts.genetic_relatedness](https://tskit.dev/tskit/docs/latest/python-api.html#tskit.TreeSequence.genetic_relatedness).

However, I imagine you're looking for the classical expected-relatedness-between-diploids quantity (as computed by SLiM). In this case, I don't have code written to do it already, but I do have:
- [individual_parents](https://github.com/tskit-dev/pyslim/blob/main/pyslim/methods.py#L531) and 
- [has_individual_parents](https://github.com/tskit-dev/pyslim/blob/main/pyslim/methods.py#L548).

I've just noticed that I don't have these listed in the docs (and will fix that shortly - so they should be at https://tskit.dev/pyslim/docs/latest/python_api.html soon) - but they give you the information you need to compute relatedness. The way I'd do this is to (a) use `ts.individual_parents` to build the sparse matrix of "who is whose parent", and (b) use some iterative algorithm to compute relatedness (I'm not up on the best way to do this).  If you figure out the right way to do it, I'd appreciate a reportback (maybe even with code?).

How's that?

**Reply by fouerghi** [reply]
Hi @petrelharp, thank you for the advice! I will try that and keep you posted.

**Reply by petrelharp** [reply]
As @bhaller notes, you have to do `treeSeqRememberIndividuals(community.allSubpopulations.individuals)` every tick in order for this to work (i.e., to make sure that you have the full pedigree of the simulation).

**Comment by bhaller** [comment]
@petrelharp, do these methods depend upon remembering every individual in the simulation?  How does remembering interact with calculating relatedness?

**Reply by petrelharp** [reply]
Ah, yes, good point - they do assume that everyone has been Remembered (and then the relatedness calculated will only be within the pedigree of the simulation, but I assume that's what we want here).

**Comment by jeromekelleher** [comment]
There are some powerful pedigree relatedness methods in [sgkit](https://pystatgen.github.io/sgkit/latest)

The [relatedness tutorial](https://pystatgen.github.io/sgkit/latest/examples/relatedness_tutorial.html) should be some help in getting started with pedigrees in sgkit.

This discussion should help with the conversion from tskit: https://github.com/tskit-dev/tskit/discussions/2711#discussioncomment-5069895

(We really should document this somewhere; I'm currently working on the sgkit paper which will hopefully raise the profile and create some interest)

**Reply by bhaller** [reply]
Very interesting @jeromekelleher.  sgkit is not on my radar at all.  How do you see the division of labor between tskit and sgkit?  I.e., a natural thing for a user (and me) to wonder is: why am I converting my tree sequence to a different form and doing this in sgkit, rather than doing this natively in tskit?

As for "we really should document this somewhere", yes!  If this is the standard/supported solution for this sort of problem going forward, we ought to have a tutorial/vignette somewhere.  SLiM manual; pyslim documentation; sgkit documentation?  Where's the right place?

**Reply by fouerghi** [reply]
Thank you Jerome! I'll make sure to check this out. It would definitely be nice to have this more streamlined for users!

**Reply by jeromekelleher** [reply]
All very good questions @bhaller! The split of functionality between tskit and sgkit is something that's not quite clear at the moment, but the basic rule of thumb will be that unless something is directly to do with trees/tree sequences it probably shouldn't be in tskit. The reasons are essentially to do with the development model - one of the key benefits of sgkit is that it's all Python (accelerated in various ways), and so it's orders of magnitude easier to add stuff in terms of developer effort.

How we package this up and communicate it to users is very much a work in progress.

**Reply by timothymillar** [reply]
Please feel free to Cc me if you have any issues or suggestions for the pedigree methods in sgkit (or the tutorial)!

**Reply by fouerghi** [reply]
Hi @timothymillar, your script to convert from the trees in tskit to input in sgkit worked perfectly! Thank you for sharing!
